spurious error on first commit after rebase
```
[info] [info] set current project to sbt_8abf1597 (in build file:/tmp/sbt_8abf1597/)
[info] [error] lmcoursier.internal.shaded.coursier.error.FetchError$DownloadingArtifacts: Error fetching artifacts:
[info] [error] file:///home/jenkins/.ivy2/local/org.scala-lang/scala-compiler/2.13.15-bin-SNAPSHOT/jars/scala-compiler.jar: wrong checksum: /home/jenkins/.ivy2/local/org.scala-lang/scala-compiler/2.13.15-bin-SNAPSHOT/jars/scala-compiler.jar (expected SHA-1 83c122d8bc099f9087008091b008e29b99e89b68 in /home/jenkins/.ivy2/local/org.scala-lang/scala-compiler/2.13.15-bin-SNAPSHOT/jars/scala-compiler.jar.sha1, got 87e3f962badf079fa40348f5e4ffa1c25fe6f85d)
[info] [error] 
[info] [error] 	at lmcoursier.internal.shaded.coursier.Artifacts$.$anonfun$fetchArtifacts$9(Artifacts.scala:365)
[info] [error] 	at lmcoursier.internal.shaded.coursier.util.Task$.$anonfun$flatMap$extension$1(Task.scala:14)
[info] [error] 	at lmcoursier.internal.shaded.coursier.util.Task$.$anonfun$flatMap$extension$1$adapted(Task.scala:14)
[info] [error] 	at lmcoursier.internal.shaded.coursier.util.Task$.wrap(Task.scala:82)
[info] [error] 	at lmcoursier.internal.shaded.coursier.util.Task$.$anonfun$flatMap$2(Task.scala:14)
[info] [error] 	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
[info] [error] 	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:51)
[info] [error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:74)
[info] [error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info] [error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info] [error] 	at java.lang.Thread.run(Thread.java:750)
[info] [error] Caused by: lmcoursier.internal.shaded.coursier.cache.ArtifactError$WrongChecksum: wrong checksum: /home/jenkins/.ivy2/local/org.scala-lang/scala-compiler/2.13.15-bin-SNAPSHOT/jars/scala-compiler.jar (expected SHA-1 83c122d8bc099f9087008091b008e29b99e89b68 in /home/jenkins/.ivy2/local/org.scala-lang/scala-compiler/2.13.15-bin-SNAPSHOT/jars/scala-compiler.jar.sha1, got 87e3f962badf079fa40348f5e4ffa1c25fe6f85d)
[info] [error] 	at lmcoursier.internal.shaded.coursier.cache.FileCache.$anonfun$validateChecksum$4(FileCache.scala:146)
[info] [error] 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
[info] [error] 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
[info] [error] 	at scala.util.Success.map(Try.scala:213)
[info] [error] 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[info] [error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:42)
[info] [error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:74)
[info] [error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info] [error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info] [error] 	at java.lang.Thread.run(Thread.java:750)
[info] [error] (update) lmcoursier.internal.shaded.coursier.error.FetchError$DownloadingArtifacts: Error fetching artifacts:
```
then
```
[error] [info] waiting for lock on /home/jenkins/.ivy2/.sbt.ivy.lock to be available...
[info] [info] compiling 1 Scala source to /tmp/sbt_8131f52f/macro-provider/target/classes ...
[info] [error] ## Exception when compiling 1 sources to /tmp/sbt_8131f52f/macro-provider/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/reflect/macros/internal/macroImpl
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.MacroImplAnnotation$lzycompute(Definitions.scala:587)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.MacroImplAnnotation(Definitions.scala:587)
[info] [error] scala.tools.nsc.typechecker.Macros.$anonfun$loadMacroImplBinding$1(Macros.scala:291)
[info] [error] scala.collection.mutable.AnyRefMap.getOrElseUpdate(AnyRefMap.scala:169)
[info] [error] scala.tools.nsc.typechecker.Macros.loadMacroImplBinding(Macros.scala:293)
[info] [error] scala.tools.nsc.typechecker.Macros.loadMacroImplBinding$(Macros.scala:289)
[info] [error] scala.tools.nsc.Global$$anon$6.loadMacroImplBinding(Global.scala:514)
[info] [error] scala.tools.nsc.typechecker.Macros.standardIsBlackbox(Macros.scala:310)
[info] [error] scala.tools.nsc.typechecker.Macros.standardIsBlackbox$(Macros.scala:308)
[info] [error] scala.tools.nsc.Global$$anon$6.standardIsBlackbox(Global.scala:514)
[info] [error] scala.tools.nsc.typechecker.AnalyzerPlugins$$anon$10.default(AnalyzerPlugins.scala:466)
[info] [error] scala.tools.nsc.typechecker.AnalyzerPlugins$$anon$10.default(AnalyzerPlugins.scala:463)
[info] [error] scala.tools.nsc.typechecker.AnalyzerPlugins.invoke(AnalyzerPlugins.scala:428)
```

--- a/src/compiler/scala/tools/nsc/Reporting.scala
+++ b/src/compiler/scala/tools/nsc/Reporting.scala
@@ -572,6 +572,7 @@ object Reporting {
         WFlagNumericWiden,
         WFlagSelfImplicit,
         WFlagUnnamedBooleanLiteral,
+        WFlagTostringInterpolated,
         WFlagValueDiscard
       = wflag()
 

--- a/src/compiler/scala/tools/nsc/settings/Warnings.scala
+++ b/src/compiler/scala/tools/nsc/settings/Warnings.scala
@@ -123,6 +123,7 @@ trait Warnings {
   val warnOctalLiteral     = BooleanSetting("-Woctal-literal", "Warn on obsolete octal syntax.") withAbbreviation "-Ywarn-octal-literal"
   val warnUnnamedBoolean   = BooleanSetting("-Wunnamed-boolean-literal", "Warn about unnamed boolean literals if there is more than one or defaults are used, unless parameter has @deprecatedName.")
   val warnUnnamedStrict    = BooleanSetting("-Wunnamed-boolean-literal-strict", "Warn about all unnamed boolean literals, unless parameter has @deprecatedName or the method has a single leading boolean parameter.").enabling(warnUnnamedBoolean :: Nil)
+  val warnToString         = BooleanSetting("-Wtostring-interpolated", "Warn when a standard interpolator uses toString.")
 
   object PerformanceWarnings extends MultiChoiceEnumeration {
     val Captured       = Choice("captured",        "Modification of var in closure causes boxing.")

--- a/src/compiler/scala/tools/reflect/FastStringInterpolator.scala
+++ b/src/compiler/scala/tools/reflect/FastStringInterpolator.scala
@@ -13,7 +13,7 @@
 package scala.tools
 package reflect
 
-import nsc.Reporting.WarningCategory.Scala3Migration
+import nsc.Reporting.WarningCategory.{Scala3Migration, WFlagTostringInterpolated}
 
 trait FastStringInterpolator extends FormatInterpolator {
   import c.universe._
@@ -101,11 +101,15 @@ trait FastStringInterpolator extends FormatInterpolator {
       val treatedContents = lit.asInstanceOf[Literal].value.stringValue
       val emptyLit = treatedContents.isEmpty
       if (i < numLits - 1) {
-        concatArgs += argsIndexed(i)
-        if (!emptyLit) concatArgs += lit
-      } else if (!emptyLit) {
-        concatArgs += lit
+        val arg = argsIndexed(i)
+        if (linting && !(arg.tpe =:= definitions.StringTpe))
+          if (arg.tpe.typeSymbol eq definitions.UnitClass)
+            runReporting.warning(arg.pos, "interpolated Unit value", WFlagTostringInterpolated, c.internal.enclosingOwner)
+          else if (!definitions.isPrimitiveValueType(arg.tpe))
+            runReporting.warning(arg.pos, "interpolation uses toString", WFlagTostringInterpolated, c.internal.enclosingOwner)
+        concatArgs += arg
       }
+      if (!emptyLit) concatArgs += lit
     }
     def mkConcat(pos: Position, lhs: Tree, rhs: Tree): Tree =
       atPos(pos)(gen.mkMethodCall(gen.mkAttributedSelect(lhs, definitions.String_+), rhs :: Nil)).setType(definitions.StringTpe)

--- a/src/compiler/scala/tools/reflect/FormatInterpolator.scala
+++ b/src/compiler/scala/tools/reflect/FormatInterpolator.scala
@@ -12,11 +12,12 @@
 
 package scala.tools.reflect
 
-import scala.reflect.macros.runtime.Context
-import scala.collection.mutable.ListBuffer
 import scala.PartialFunction.cond
-import scala.util.chaining._
+import scala.collection.mutable.ListBuffer
+import scala.reflect.macros.runtime.Context
+import scala.tools.nsc.Reporting.WarningCategory, WarningCategory.WFlagTostringInterpolated
 import scala.util.matching.Regex.Match
+import scala.util.chaining._
 
 import java.util.Formattable
 
@@ -30,6 +31,14 @@ abstract class FormatInterpolator {
   import c.universe.{Match => _, _}
   import definitions._
   import treeInfo.Applied
+
+  protected var linting = settings.warnToString.value
+
+  protected final def withoutLinting[A](body: => A): A = {
+    val linted = linting
+    linting = false
+    try body finally linting = linted
+  }
 
   private def bail(msg: String) = global.abort(msg)
 
@@ -87,8 +96,9 @@ abstract class FormatInterpolator {
 
     def argType(argi: Int, types: Type*): Type = {
       val tpe = argTypes(argi)
-      types.find(t => argConformsTo(argi, tpe, t))
-        .orElse(types.find(t => argConvertsTo(argi, tpe, t)))
+      types.find(t => t != AnyTpe && argConformsTo(argi, tpe, t))
+        .orElse(types.find(t => t != AnyTpe && argConvertsTo(argi, tpe, t)))
+        .orElse(types.find(t => t == AnyTpe && argConformsTo(argi, tpe, t)))
         .getOrElse {
           val msg = "type mismatch" + {
             val req = raw"required: (.*)".r.unanchored
@@ -120,7 +130,7 @@ abstract class FormatInterpolator {
     // Check the % fields in this part.
     def loop(remaining: List[Tree], n: Int): Unit =
       remaining match {
-        case part0 :: more =>
+        case part0 :: remaining =>
           val part1 = part0 match {
             case Literal(Constant(x: String)) => x
             case _ => throw new IllegalArgumentException("internal error: argument parts must be a list of string literals")
@@ -130,14 +140,17 @@ abstract class FormatInterpolator {
 
           def insertStringConversion(): Unit = {
             amended += "%s" + part
-            convert += Conversion(formatPattern.findAllMatchIn("%s").next(), part0.pos, argc)  // improve
-            argType(n-1, AnyTpe)
+            val cv = Conversion(part0.pos, argc)
+            cv.accepts(argType(n-1, AnyTpe))
+            convert += cv
+            cv.lintToString(argTypes(n-1))
           }
           def errorLeading(op: Conversion) = op.errorAt(Spec)(s"conversions must follow a splice; ${Conversion.literalHelp}")
           def accept(op: Conversion): Unit = {
             if (!op.isLeading) errorLeading(op)
             op.accepts(argType(n-1, op.acceptableVariants: _*))
             amended += part
+            op.lintToString(argTypes(n-1))
           }
 
           if (n == 0) amended += part
@@ -164,7 +177,7 @@ abstract class FormatInterpolator {
             else if (!cv.isLiteral && !cv.isIndexed) errorLeading(cv)
             formatting = true
           }
-          loop(more, n = n + 1)
+          loop(remaining, n = n + 1)
         case Nil =>
       }
     loop(parts, n = 0)
@@ -178,7 +191,11 @@ abstract class FormatInterpolator {
     val format = amended.mkString
     if (actuals.isEmpty && !formatting) constantly(format)
     else if (!reported && actuals.forall(treeInfo.isLiteralString)) constantly(format.format(actuals.map(_.asInstanceOf[Literal].value.value).toIndexedSeq: _*))
-    else if (!formatting) concatenate(amended.map(p => constantly(p.stripPrefix("%s"))).toList, actuals.toList)
+    else if (!formatting) {
+      withoutLinting { // already warned
+        concatenate(amended.map(p => constantly(p.stripPrefix("%s"))).toList, actuals.toList)
+      }
+    }
     else {
       val scalaPackage = Select(Ident(nme.ROOTPKG), TermName("scala"))
       val newStringOps = Select(
@@ -218,6 +235,7 @@ abstract class FormatInterpolator {
         case ErrorXn                     => op(0)
         case DateTimeXn if op.length > 1 => op(1)
         case DateTimeXn                  => '?'
+        case StringXn if op.isEmpty      => 's' // accommodate the default %s
         case _ => op(0)
       }
 
@@ -295,13 +313,19 @@ abstract class FormatInterpolator {
           }
         case _ => true
       }
+    def lintToString(arg: Type): Unit =
+      if (linting && kind == StringXn && !(arg =:= StringTpe))
+        if (arg.typeSymbol eq UnitClass)
+          warningAt(CC)("interpolated Unit value", WFlagTostringInterpolated)
+        else if (!definitions.isPrimitiveValueType(arg))
+          warningAt(CC)("interpolation uses toString", WFlagTostringInterpolated)
 
     // what arg type if any does the conversion accept
     def acceptableVariants: List[Type] =
       kind match {
         case StringXn if hasFlag('#') => FormattableTpe :: Nil
         case StringXn                 => AnyTpe :: Nil
-        case BooleanXn                => BooleanTpe :: NullTpe :: Nil
+        case BooleanXn                => BooleanTpe :: NullTpe :: AnyTpe :: Nil // warn if not boolean
         case HashXn                   => AnyTpe :: Nil
         case CharacterXn              => CharTpe :: ByteTpe :: ShortTpe :: IntTpe :: Nil
         case IntegralXn               => IntTpe :: LongTpe :: ByteTpe :: ShortTpe :: BigIntTpe :: Nil
@@ -331,7 +355,7 @@ abstract class FormatInterpolator {
 
     def groupPosAt(g: SpecGroup, i: Int) = pos.withPoint(pos.point + descriptor.offset(g, i))
     def errorAt(g: SpecGroup, i: Int = 0)(msg: String)   = c.error(groupPosAt(g, i), msg).tap(_ => reported = true)
-    def warningAt(g: SpecGroup, i: Int = 0)(msg: String) = c.warning(groupPosAt(g, i), msg)
+    def warningAt(g: SpecGroup, i: Int = 0)(msg: String, cat: WarningCategory = WarningCategory.Other) = c.callsiteTyper.context.warning(groupPosAt(g, i), msg, cat, Nil)
   }
 
   object Conversion {
@@ -356,6 +380,9 @@ abstract class FormatInterpolator {
         case None     => new Conversion(m, p, ErrorXn, argc).tap(_.errorAt(Spec)(s"Missing conversion operator in '${m.matched}'; $literalHelp"))
       }
     }
+    // construct a default %s conversion
+    def apply(p: Position, argc: Int): Conversion =
+      new Conversion(formatPattern.findAllMatchIn("%").next(), p, StringXn, argc)
     val literalHelp = "use %% for literal %, %n for newline"
   }
 

--- a/test/files/neg/stringinterpolation_macro-neg.check
+++ b/test/files/neg/stringinterpolation_macro-neg.check
@@ -10,11 +10,6 @@ stringinterpolation_macro-neg.scala:15: error: too many arguments for interpolat
 stringinterpolation_macro-neg.scala:16: error: too few arguments for interpolated string
   new StringContext("", "").f()
                              ^
-stringinterpolation_macro-neg.scala:19: error: type mismatch;
- found   : String
- required: Boolean, Null
-  f"$s%b"
-     ^
 stringinterpolation_macro-neg.scala:20: error: type mismatch;
  found   : String
  required: Char, Byte, Short, Int
@@ -162,6 +157,9 @@ stringinterpolation_macro-neg.scala:77: error: Missing conversion operator in '%
 stringinterpolation_macro-neg.scala:80: error: conversions must follow a splice; use %% for literal %, %n for newline
   f"${d}random-leading-junk%d"
                            ^
+stringinterpolation_macro-neg.scala:19: warning: Boolean format is null test for non-Boolean
+  f"$s%b"
+       ^
 stringinterpolation_macro-neg.scala:63: warning: Index is not this arg
   f"${8}%d ${9}%1$$d"
                 ^
@@ -171,5 +169,5 @@ stringinterpolation_macro-neg.scala:64: warning: Argument index ignored if '<' f
 stringinterpolation_macro-neg.scala:65: warning: Index is not this arg
   f"$s%s $s%1$$s"
             ^
-3 warnings
-46 errors
+4 warnings
+45 errors

--- a/test/files/neg/tostring-interpolated.check
+++ b/test/files/neg/tostring-interpolated.check
@@ -1,0 +1,41 @@
+tostring-interpolated.scala:7: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.f
+  def f = f"$c" // warn
+             ^
+tostring-interpolated.scala:8: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.s
+  def s = s"$c" // warn
+             ^
+tostring-interpolated.scala:9: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.r
+  def r = raw"$c" // warn
+               ^
+tostring-interpolated.scala:11: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.format
+  def format = f"${c.x}%d in $c or $c%s" // warn using c.toString // warn
+                              ^
+tostring-interpolated.scala:11: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.format
+  def format = f"${c.x}%d in $c or $c%s" // warn using c.toString // warn
+                                      ^
+tostring-interpolated.scala:13: warning: Boolean format is null test for non-Boolean
+  def bool = f"$c%b" // warn just a null check
+                  ^
+tostring-interpolated.scala:15: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.oops
+  def oops = s"${null} slipped thru my fingers" // warn
+                 ^
+tostring-interpolated.scala:20: error: interpolation uses toString
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=T.greeting
+  def greeting = s"$sb, world" // warn
+                    ^
+tostring-interpolated.scala:31: error: interpolated Unit value
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=Mitigations.unitized
+  def unitized = s"unfortunately $shown" // warn accidental unit value
+                                  ^
+tostring-interpolated.scala:32: error: interpolated Unit value
+Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=w-flag-tostring-interpolated, site=Mitigations.funitized
+  def funitized = f"unfortunately $shown" // warn accidental unit value
+                                       ^
+1 warning
+9 errors

--- a/test/files/neg/tostring-interpolated.scala
+++ b/test/files/neg/tostring-interpolated.scala
@@ -1,0 +1,36 @@
+//> using options -Wconf:cat=w-flag-tostring-interpolated:e -Wtostring-interpolated
+
+case class C(x: Int)
+
+trait T {
+  def c = C(42)
+  def f = f"$c" // warn
+  def s = s"$c" // warn
+  def r = raw"$c" // warn
+
+  def format = f"${c.x}%d in $c or $c%s" // warn using c.toString // warn
+
+  def bool = f"$c%b" // warn just a null check
+
+  def oops = s"${null} slipped thru my fingers" // warn
+
+  def ok = s"${c.toString}"
+
+  def sb = new StringBuilder().append("hello")
+  def greeting = s"$sb, world" // warn
+}
+
+class Mitigations {
+
+  val s = "hello, world"
+  val i = 42
+  def shown() = println("shown")
+
+  def ok = s"$s is ok"
+  def jersey = s"number $i"
+  def unitized = s"unfortunately $shown" // warn accidental unit value
+  def funitized = f"unfortunately $shown" // warn accidental unit value
+
+  def nopct = f"$s is ok"
+  def nofmt = f"number $i"
+}

--- a/test/files/run/f-interpolator-unit.scala
+++ b/test/files/run/f-interpolator-unit.scala
@@ -33,7 +33,7 @@ object Test extends App {
   final val tester = "hello"
   final val number = "42"  // strings only, alas
 
-  def assertEquals(s0: String, s1: String) = assert(s0 == s1, s"$s0 == $s1")
+  def assertEquals(s0: String, s1: String, i: Int = -1) = assert(s0 == s1, s"$s0 == $s1${if (i >= 0) " at " + i.toString else ""}")
 
   def noEscape() = {
     val s = "string"
@@ -134,7 +134,7 @@ object Test extends App {
       f"${null}%b"  -> "false",
       f"${false}%b" -> "false",
       f"${true}%b"  -> "true",
-      f"${true && false}%b"                 -> "false",
+      f"${true && false}%b"                     -> "false",
       f"${java.lang.Boolean.valueOf(false)}%b"  -> "false",
       f"${java.lang.Boolean.valueOf(true)}%b"   -> "true",
 
@@ -255,7 +255,7 @@ object Test extends App {
       f"z" -> "z"
     )
 
-    for ((f, s) <- ss) assertEquals(s, f)
+    for (((f, s), i) <- ss.zipWithIndex) assertEquals(s, f, i)
   }
 
   noEscape()

